src/Jellyfin.LiveTv/Listings/ListingsManager.cs

```diff
@@ -0,0 +1,470 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+using Jellyfin.LiveTv.Configuration;
+using Jellyfin.LiveTv.EmbyTV;
+using Jellyfin.LiveTv.Guide;
+using MediaBrowser.Common.Configuration;
+using MediaBrowser.Common.Extensions;
+using MediaBrowser.Controller.LiveTv;
+using MediaBrowser.Model.Dto;
+using MediaBrowser.Model.LiveTv;
+using MediaBrowser.Model.Tasks;
+using Microsoft.Extensions.Logging;
+
+namespace Jellyfin.LiveTv.Listings;
+
+/// <inheritdoc />
+public class ListingsManager : IListingsManager
+{
+    private readonly ILogger<ListingsManager> _logger;
+    private readonly IConfigurationManager _config;
+    private readonly ITaskManager _taskManager;
+    private readonly ITunerHostManager _tunerHostManager;
+    private readonly IListingsProvider[] _listingsProviders;
+
+    private readonly ConcurrentDictionary<string, EpgChannelData> _epgChannels = new(StringComparer.OrdinalIgnoreCase);
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="ListingsManager"/> class.
+    /// </summary>
+    /// <param name="logger">The <see cref="ILogger{TCategoryName}"/>.</param>
+    /// <param name="config">The <see cref="IConfigurationManager"/>.</param>
+    /// <param name="taskManager">The <see cref="ITaskManager"/>.</param>
+    /// <param name="tunerHostManager">The <see cref="ITunerHostManager"/>.</param>
+    /// <param name="listingsProviders">The <see cref="IListingsProvider"/>.</param>
+    public ListingsManager(
+        ILogger<ListingsManager> logger,
+        IConfigurationManager config,
+        ITaskManager taskManager,
+        ITunerHostManager tunerHostManager,
+        IEnumerable<IListingsProvider> listingsProviders)
+    {
+        _logger = logger;
+        _config = config;
+        _taskManager = taskManager;
+        _tunerHostManager = tunerHostManager;
+        _listingsProviders = listingsProviders.ToArray();
+    }
+
+    /// <inheritdoc />
+    public async Task<ListingsProviderInfo> SaveListingProvider(ListingsProviderInfo info, bool validateLogin, bool validateListings)
+    {
+        ArgumentNullException.ThrowIfNull(info, nameof(info));
+
+        // Hack to make the object a pure ListingsProviderInfo instead of an AddListingProvider
+        // ServerConfiguration.SaveConfiguration crashes during xml serialization for AddListingProvider
+        info = JsonSerializer.Deserialize<ListingsProviderInfo>(JsonSerializer.SerializeToUtf8Bytes(info))!;
```

Comment:
Can this be removed? It looks like this is only called by the API endpoint, which takes a `ListingsProviderInfo` already.